### DISCUSSION TITLE
How/can you test Haraka locally

### DISCUSSION THREAD

**Discussion by nogginbox** [haraka/Haraka#3190] [Q&A] [answered]
I'd like to test a docker container of Haraka running locally with my required config before I attempt to get it running on a server and update all the MX records of my domain. I don't know enough about mail protocols to know how I'd do this.

My current plan is to test it on a live server but setup mx records for a test subdomain before I move it to use my real domain. Is there a better way?

**Comment by msimerson** [comment]
Easiest solution I can think of it run the Haraka instance exactly as normal except use a custom DNS server to lie to Haraka about the world. Email relies heavily on DNS so this might take a bit of trial and error to get all the records in place, but it would work and also serve as a dev/test environment before future deployments.

**Comment by nogginbox** [comment, marked as answer]
By learning some simple SMTP commands I found I could test it locally using telnet. Here is a simple example of sending an email to my server:
```
telnet localhost 25
> 220 localhost v9.0.3.0 ESMTP ready

ehlo sender.co
> 250-localhost Hello sender.co, haven't we met before?

mail from:<test@sending-domain.co>
> 250 Ok

rcpt to:<us@my-domain.co>
> 250 Ok

data
> 354 end with <CRLF>.<CRLF>

Subject: Test Email
Date: Tue, 18 Jul 2023 21:41:00
From: test@docker.localhost
To: someone.awesome@alias-domain.co
Message-Id: <56d4ae71-e38c-4530-81ea-60a87546e9e7@docker.local>

Test content as descibed in the RFC http://www.faqs.org/rfcs/rfc2822.html.
.
```

**Reply by msimerson** [reply]
You can also test locally against the TLS versions of the port using openssl:

`openssl s_client -connect locahost:465`
